src/ccs.rs

```diff
@@ -0,0 +1,636 @@
+  constants::{BN_LIMB_WIDTH, BN_N_LIMBS, NUM_FE_FOR_RO, NUM_HASH_BITS},
+  errors::NovaError,
+  gadgets::{
+    nonnative::{bignat::nat_to_limbs, util::f_to_nat},
+    utils::scalar_as_base,
+  },
+  r1cs::{R1CSInstance, R1CSShape, R1CSWitness, R1CS},
+  traits::{
+    commitment::CommitmentEngineTrait, commitment::CommitmentTrait, AbsorbInROTrait, Group, ROTrait,
+  },
+  utils::*,
+  Commitment, CommitmentKey, CE,
+};
+use bitvec::vec;
+use core::{cmp::max, marker::PhantomData};
+use ff::Field;
+use flate2::{write::ZlibEncoder, Compression};
+use itertools::concat;
+use rayon::prelude::*;
+use serde::{Deserialize, Serialize};
+use sha3::{Digest, Sha3_256};
+use std::ops::{Add, Mul};
+
+// TODO: Committed CCS using MLE (see src/spartan/pp.rs)
+// TODO: Linearized CCS struct and methods, separate struct similar to RelaxedR1CS
+
+/// Public parameters for a given CCS
+#[derive(Clone, Serialize, Deserialize)]
+#[serde(bound = "")]
+pub struct CCS<G: Group> {
+  _p: PhantomData<G>,
+}
+
+/// A type that holds the shape of a CCS instance
+/// Unlike R1CS we have a list of matrices M instead of only A, B, C
+/// We also have t, q, d constants and c (vector), S (set)
+/// As well as m, n, s, s_prime
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(bound = "")]
+pub struct CCSShape<G: Group> {
+  pub(crate) num_cons: usize,
+  pub(crate) num_vars: usize,
+  pub(crate) num_io: usize,
+
+  pub(crate) M: Vec<SparseMatrix<G>>,
+  pub(crate) t: usize,
+  pub(crate) q: usize,
+  pub(crate) d: usize,
+  pub(crate) S: Vec<Vec<usize>>,
+
+  // Was: usize
+  pub(crate) c: Vec<G::Scalar>,
+
+  // m is the number of columns in M_i
+  pub(crate) m: usize,
+  // n is the number of rows in M_i
+  pub(crate) n: usize,
+  // s = log m
+  pub(crate) s: usize,
+  // s_prime = log n
+  pub(crate) s_prime: usize,
+}
+
+/// A type that holds a witness for a given CCS instance
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct CCSWitness<G: Group> {
+  // Vector W in F^{n - l - 1}
+  W: Vec<G::Scalar>,
+}
+
+// TODO: Make sure this is in the right form for committed CCS using MLE, possibly a separate type
+/// A type that holds an CCS instance
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(bound = "")]
+pub struct CCSInstance<G: Group> {
+  // XXX: Move commitment out of CCSInstance for more clean conceptual separation?
+  // (Pedersen) Commitment to a witness
+  pub(crate) comm_W: Commitment<G>,
+
+  // Public input x in F^l
+  pub(crate) X: Vec<G::Scalar>,
+}
+
+/// A type that holds the shape of a Committed CCS (CCCS) instance
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(bound = "")]
+pub struct CCCSShape<G: Group> {
+  // Sequence of sparse MLE polynomials in s+s' variables M_MLE1, ..., M_MLEt
+  // TODO This should be MLE, but possible we don't have to keep in struct?
+  // Exists in paper but not multifolding-poc
+  pub(crate) M_MLE: Vec<SparseMatrix<G>>,
```

Comment:
We already have a type that matches perfectly the MLE of the matrix.

Note that for https://github.com/privacy-scaling-explorations/multifolding-poc/blob/main/src/util/mle.rs#L30-L41
We get back a `DenseMultilinearExtension<F>` which is exactly the same as: https://github.com/privacy-scaling-explorations/Nova/blob/hypernova/src/spartan/polynomial.rs#L80-L83 and contains also the same methods needed.